src/attr_file.c

```diff
@@ -397,7 +397,7 @@ bool git_attr_fnmatch__match(
 
 		/* for attribute checks or root ignore checks, fail match */
 		if (!(match->flags & GIT_ATTR_FNMATCH_IGNORE) ||
-			path->basename == path->path)
+			path->basename == relpath)
```

Comment:
I honestly don't recall if I was intending for this to be pointer comparison, however I added some debug logging this morning and it does work, because the pointers in question are all derived from offsets into `path->path` (`basename` gets derived in `git_attr_path__init`; `relpath` gets derived earlier in `git_attr_fnmatch__match`).

I would _definitely_ like someone to triple-check my thought process on this though. I don't know if this is actually helpful to others, but it helps my thinking to write it down so:

Based on the immediate closing `if`-statement: `(match->flags & GIT_ATTR_FNMATCH_DIRECTORY) && !path->is_dir`), my read of the original early return is:

a) IF we're trying to match a directory
b) AND the path we're looking at is _not_ a directory
c) AND the `basename` and `path` are the same (i.e., they represent an entry at the root of the tree),
THEN we can't possibly match because we must be looking at a *file* (because of `b)` ) in the root of the git tree.

The change to use `relpath` is an attempt to say the same thing, but to extend `c)` to include any file directly under the `containing_dir` of the rule we're comparing with.

Or at least that's what I've been telling myself. :)

I'll throw some more test cases at it. I'm still not 100% convinced that replacing `path->path` with `relpath` is the perfect solution. (Does it work if the .gitignore is at the root, and it's ignoring directories in a subdirectory? I think it doesn't...) I think it does preserve existing behavior w.r.t. files at the root, and extends it to files at the root of directories w/ .gitignores.